risc0/zkvm/src/host/api/mod.rs

```diff
@@ -319,28 +289,45 @@ pub enum AssetRequest {
     Path(PathBuf),
 }
 
+/// Provides information about the result of execution.
+pub struct SessionInfo {
+    /// The number of segments.
+    pub segments: u32,
+
+    /// The data publicly committed by the guest program.
+    pub journal: Bytes,
+
+    /// The [ExitCode] of the session.
+    pub exit_code: ExitCode,
```

Comment:
Instead of journal and exit_code, it'd be nice to have this be `ReceiptMetadata`, which essentially represents the set of public claims about an execution. For composition, being able to get the `ReceiptMetadata` for an execution is very useful because it allows starting execution and proving for one program as soon as all of it's dependencies have been executed, rather than waiting for dependencies to be proven.

One downside to this would be pushing the `journal` one layer down, which is the main field a developer is likely interested in. One option would be to have these fields be top-level in `SessionInfo` and have a `get_metadata` method that constructs the `ReceiptMetadata` struct if needed.

Reply:
This is a really good point, I like the idea of having a `get_metadata`. I'll play around with this.

Reply:
I like this idea, but it turns out that the executor doesn't yet have the smarts to construct a `ReceiptMetadata`. Also, I don't know how we can construct one without running the entire program since only the final segment is guaranteed to have outputs and the final output digest.